### PR TITLE
CSI-PowerStore Resiliency Integration & Scalability Tests

### PR DESCRIPTION
**Pull request by suryagupta4**
# Description
This PR is for making the changes wrt to Integration and Scalability tests for csi-powerstore driver

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/587 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Ran integration tests for iSCSI
iscsi-driver-pod
iscsi-interface-down
iscsi-kubeletdown
iscsi-multi-nw
iscsi-reboot

-----------------

- [x] Ran integration tests for NVMe-tcp
nvme-driver-pods
nvme-interface-down
nvme-kubelet-down
nvme-multi-nw
nvme-reboot

------------------

- [x] Ran integration tests for NFS
nfs-driver-pod
nfs-interface-down
nfs-kubelet-down
nfs-multi-nw
nfs-reboot

### CODE REVIEW

File: test/podmontest/insips.sh [hunk 1]
```diff
@@ -0,0 +1,109 @@
+#!/bin
```
Comment (alikdell) [inline test/podmontest/insips.sh:1]: File name with current standard naming should be "insps.sh"

File: internal/monitor/features/integration.feature [hunk 2]
```diff
@@ -132,6 +145,25 @@ Feature: Integration Test
...
+    Then validate that all pods are running within <runSecs> seconds
+    And labeled pods are on a different node
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass  | workers     | primary | failure         | failSecs | deploySecs | runSecs | nodeCleanSecs |
+     # Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "pstore"    | "powerstore-nfs" | "one-third" | "zero"  | "interfacedown" | 120      | 600        | 600     | 600           |
+      | ""         | "3-5"       | "2-2" | "0-0" | "pstore"    | "powerstore-nfs" | "one-third" | "zero"  | "interfacedown" | 240      | 600        | 600     | 600           |
```
Comment (alikdell) [inline internal/monitor/features/integration.feature:167]: We should add iscsi and nvme type too

File: internal/monitor/features/integration.feature [hunk 3]
```diff
@@ -220,6 +252,26 @@ Feature: Integration Test
...
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass | workers     | primary | failure       | failSecs | deploySecs | runSecs | nodeCleanSecs |
+      # Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "pstore"    | "powerstore-nfs"  | "one-third" | "zero"  | "kubeletdown" | 600      | 900        | 900     | 900           |
+      # Slightly more pods, increasing number of vols and devs
+      | ""         | "3-5"       | "1-1" | "0-0" | "pstore"    | "powerstore-nfs"  | "one-third" | "zero"  | "kubeletdown" | 600      | 900        | 900     | 900           |
+
```
Comment (alikdell) [inline internal/monitor/features/integration.feature:274]: We should add iscsi and nvme type too

File: internal/monitor/features/integration.feature [hunk 4]
```diff
@@ -132,6 +145,25 @@ Feature: Integration Test
...
+    Then validate that all pods are running within <runSecs> seconds
+    And labeled pods are on a different node
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass  | workers     | primary | failure         | failSecs | deploySecs | runSecs | nodeCleanSecs |
+     # Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "pstore"    | "powerstore-nfs" | "one-third" | "zero"  | "interfacedown" | 120      | 600        | 600     | 600           |
+      | ""         | "3-5"       | "2-2" | "0-0" | "pstore"    | "powerstore-nfs" | "one-third" | "zero"  | "interfacedown" | 240      | 600        | 600     | 600           |
```
Comment (alikdell) [inline internal/monitor/features/integration.feature:166]: We should add iscsi and nvme type too

File: internal/monitor/features/integration.feature [hunk 5]
```diff
@@ -364,6 +416,23 @@ Feature: Integration Test
...
+    And <podsPerNode> pods per node with <nVol> volumes and <nDev> devices using <driverType> and <storageClass> in <deploySecs>
+    Then validate that all pods are running within <deploySecs> seconds
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+    And <podsPerNode> pods per node with <nVol> volumes and <nDev> devices using <driverType> and <storageClass> in <deploySecs>
+    Then validate that all pods are running within <deploySecs> seconds
+    Then finally cleanup everything
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass  | workers     | primary | failure         | failSecs | deploySecs | nodeCleanSecs |
+      | ""         | "1-2"       | "1-1" | "0-0" | "pstore"   | "powerstore-nfs"      | "one-third" | "zero"  | "interfacedown" | 600      | 900        | 900           |
```
Comment (alikdell) [inline internal/monitor/features/integration.feature:436]: We should add iscsi and nvme type too

File: internal/monitor/features/integration.feature [hunk 6]
```diff
@@ -506,6 +575,23 @@ Feature: Integration Test
...
+    Then validate that all pods are running within <deploySecs> seconds
+    When I fail <workers> worker nodes and <primary> primary nodes with <failure> failure for <failSecs> seconds
+    Then validate that all pods are running within <runSecs> seconds
+    And labeled pods are on a different node
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass | workers     | primary | failure         | failSecs | deploySecs | runSecs | nodeCleanSecs |
+      | ""         | "1-1"       | "1-1" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "interfacedown" | 120      | 240        | 300     | 300           |
```
Comment (alikdell) [inline internal/monitor/features/integration.feature:595]: We should add iscsi and nvme type too

File: internal/monitor/features/integration.feature [hunk 7]
```diff
@@ -621,6 +707,29 @@ Feature: Integration Test
...
+       #Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 240      | 600        | 600     | 600          |
+   #   | ""         | "1-2"       | "2-2" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 600      | 900        | 900     | 900           |
+   #   | ""         | "1-2"       | "4-4" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 120      | 240        | 300     | 600           |
+      # Slightly more pods, increasing number of vols and devs
+      | ""         | "3-5"       | "1-1" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 240      | 900        | 900     | 900           |
+    # | ""         | "3-5"       | "2-2" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 240      | 240        | 300     | 600           |
+    # | ""         | "5-10"       | "1-1" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 1200      | 2000       | 2000     | 2000           |
```
Comment (alikdell) [inline internal/monitor/features/integration.feature:732]: We should add iscsi and nvme type too
And delete commented tests

File: internal/monitor/features/integration.feature [hunk 8]
```diff
@@ -621,6 +707,29 @@ Feature: Integration Test
...
+    When I fail <workers> worker nodes and <primary> primary nodes with <failure> failure for <failSecs> seconds
+    Then validate that all pods are running within <runSecs> seconds
+    And labeled pods are on a different node
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass | workers     | primary | failure  | failSecs | deploySecs | runSecs | nodeCleanSecs |
+       #Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 240      | 600        | 600     | 600          |
+   #   | ""         | "1-2"       | "2-2" | "0-0" | "pstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 600      | 900        | 900     | 900           |
```
Comment (alikdell) [inline internal/monitor/features/integration.feature:727]: Delete if not required, don't have keep commented test

File: test/sh/scaleup-powerstore.sh [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+CWD=$(pwd)
+NVOLUMES=2
+STORAGECLASS=powerstore-nfs
+MAXPODS=100
```
Comment (alikdell) [inline test/sh/scaleup-powerstore.sh:18]: Should be 200 or more

File: test/sh/scaleup-powerstore.sh [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+fi
+
+BOUNCEIPTIME=1200
+instances="108"
+if [ $instances -le $MAXPODS ]; then
+cd ../podmontest; sh insps.sh --instances "$instances" --nvolumes $NVOLUMES --storage-class $STORAGECLASS; cd $CWD
+wait_on_running
+sh ../sh/nway.sh --ns pstore --bounceipseconds $BOUNCEIPTIME --maxiterations 12  --timeoutseconds 1500
+fi
+
```
Comment (alikdell) [inline test/sh/scaleup-powerstore.sh:101]: We should scale up to 200 pods atleast, currently it is ended at 108

File: test/podmontest/insips.sh (hunk 1 above)
Comment (suryagupta4) [reply test/podmontest/insips.sh]: Renamed the file to insps.sh

File: internal/monitor/features/integration.feature (hunk 2 above)
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:173]: Added iscsi & nvme types.

File: internal/monitor/features/integration.feature (hunk 3 above)
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:288]: Added iscsi & nvme types.

File: internal/monitor/features/integration.feature (hunk 5 above)
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:452]: Added iscsi & nvme types.

File: internal/monitor/features/integration.feature (hunk 6 above)
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:613]: Added iscsi & nvme types.

File: internal/monitor/features/integration.feature (hunk 7 above)
Comment (suryagupta4) [reply internal/monitor/features/integration.feature]: Added iscsi & nvme types.
Removed commented lines.

File: internal/monitor/features/integration.feature (hunk 8 above)
Comment (suryagupta4) [reply internal/monitor/features/integration.feature]: Added iscsi & nvme types.
Removed commented lines.

File: test/podmontest/deploy/values-powerstore-nfs.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
+podmonTest:
```
[the remark this answers is not in the document]
Comment (suryagupta4) [reply test/podmontest/deploy/values-powerstore-nfs.yaml:1]: Added iscsi & nvme test files.

**Comment by alikdell** [comment]
In above section "How Has This Been Tested?" is wrong. Changes in this PR is for integration testing and Scalability test, please provide logs for integration test and Scalability test. Since complete logs might have internal IPs etc, add the last few statements from both test.

**Review by alikdell — DISMISSED** [review]
Demo on integration and Scalability testing

**Comment by suryagupta4** [comment]
> In above section "How Has This Been Tested?" is wrong. Changes in this PR is for integration testing and Scalability test, please provide logs for integration test and Scalability test. Since complete logs might have internal IPs etc, add the last few statements from both test.

Added screenshots for integration tests ran for iSCSI, NVMe-tcp & NFS protocol

File: internal/monitor/features/integration.feature [hunk 12]
```diff
@@ -657,3 +810,23 @@ Feature: Integration Test
...
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore"  | "powerstore-nfs"      | "one-third" | "zero"  | "driverpod" | "offline.powerstore.storage.dell.com" | 120      | 300        | 300     | 600           | 
+      | ""         | "1-3"       | "2-2" | "0-0" | "powerstore"  | "powerstore-nfs"      | "one-third" | "zero"  | "driverpod" | "offline.powerstore.storage.dell.com" | 120      | 300        | 300     | 600           |
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore"  | "powerstore-iscsi"    | "one-third" | "zero"  | "driverpod" | "offline.powerstore.storage.dell.com" | 120      | 300        | 300     | 600           | 
+      | ""         | "1-3"       | "2-2" | "0-0" | "powerstore"  | "powerstore-iscsi"    | "one-third" | "zero"  | "driverpod" | "offline.powerstore.storage.dell.com" | 120      | 300        | 300     | 600           |
+      #| ""         | "1-2"       | "1-1" | "0-0" | "powerstore"  | "powerstore-nvmetcp"  | "one-third" | "zero"  | "driverpod" | "offline.powerstore.storage.dell.com" | 120      | 300        | 300     | 600           |
```
Comment (shaynafinocchiaro) [inline internal/monitor/features/integration.feature:831]: same as above

File: test/podmontest/deploy/values-powerstore-nfs.yaml [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  image: "$REGISTRY_HOST:$REGISTRY_PORT//podmontest:v0.0.54"
```
Comment (shaynafinocchiaro) [inline test/podmontest/deploy/values-powerstore-nfs.yaml:2]: Reference to tester image has additional slash

File: test/podmontest/deploy/values-powerstore-nvme.yaml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+podmonTest:
+  image: "$REGISTRY_HOST:$REGISTRY_PORT//podmontest:v0.0.54"
```
Comment (shaynafinocchiaro) [inline test/podmontest/deploy/values-powerstore-nvme.yaml:2]: Reference to tester image has additional slash

File: test/podmontest/insps.sh [hunk 15]
```diff
@@ -0,0 +1,109 @@
+#!/bin
+# Copyright (c) 2022 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (shaynafinocchiaro) [inline test/podmontest/insps.sh:2]: please update copyright year

File: test/sh/scaleup-powerstore.sh [hunk 16]
```diff
@@ -0,0 +1,102 @@
+#!/bin/sh
+# Copyright (c) 2022 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (shaynafinocchiaro) [inline test/sh/scaleup-powerstore.sh:2]: please update copyright year

File: internal/monitor/features/integration.feature [hunk 17]
```diff
@@ -220,6 +258,34 @@ Feature: Integration Test
...
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore"    | "powerstore-nfs"  | "one-third" | "zero"  | "kubeletdown" | 600      | 900        | 900     | 900           |
+      # Slightly more pods, increasing number of vols and devs
+      | ""         | "3-5"       | "1-1" | "0-0" | "powerstore"    | "powerstore-nfs"  | "one-third" | "zero"  | "kubeletdown" | 600      | 900        | 900     | 900           |
+      # Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore"    | "powerstore-iscsi"  | "one-third" | "zero"  | "kubeletdown" | 600      | 900        | 900     | 900           |
+      # Slightly more pods, increasing number of vols and devs
+      | ""         | "3-5"       | "1-1" | "0-0" | "powerstore"    | "powerstore-iscsi"  | "one-third" | "zero"  | "kubeletdown" | 600      | 900        | 900     | 900           |
+      # Small number of pods, increasing number of vols and devs
+      #| ""         | "1-2"       | "1-1" | "0-0" | "powerstore"    | "powerstore-nvmetcp"  | "one-third" | "zero"  | "kubeletdown" | 600      | 900        | 900     | 900           |
```
[the remark this answers is not in the document]
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:285]: As we have to run e2e & longevity tests via csm-pipeline which doesn't support nvme-tcp, we have commented this for now. Manual testing with nvme-tcp has been done.

File: internal/monitor/features/integration.feature [hunk 18]
```diff
@@ -364,6 +430,25 @@ Feature: Integration Test
...
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+    And <podsPerNode> pods per node with <nVol> volumes and <nDev> devices using <driverType> and <storageClass> in <deploySecs>
+    Then validate that all pods are running within <deploySecs> seconds
+    Then finally cleanup everything
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass  | workers     | primary | failure         | failSecs | deploySecs | nodeCleanSecs |
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore"   | "powerstore-nfs"      | "one-third" | "zero"  | "interfacedown" | 600      | 900        | 900           |
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore"   | "powerstore-iscsi"      | "one-third" | "zero"  | "interfacedown" | 600      | 900        | 900           |
+      #| ""         | "1-2"       | "1-1" | "0-0" | "powerstore"   | "powerstore-nvmetcp"      | "one-third" | "zero"  | "interfacedown" | 600      | 900        | 900           |
```
[the remark this answers is not in the document]
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:451]: As we have to run e2e & longevity tests via csm-pipeline which doesn't support nvme-tcp, we have commented this for now. Manual testing with nvme-tcp has been done.

File: internal/monitor/features/integration.feature [hunk 19]
```diff
@@ -506,6 +591,25 @@ Feature: Integration Test
...
+    Then validate that all pods are running within <runSecs> seconds
+    And labeled pods are on a different node
+    And the taints for the failed nodes are removed within <nodeCleanSecs> seconds
+    Then finally cleanup everything
+
+    Examples:
+      | kubeConfig | podsPerNode | nVol  | nDev  | driverType | storageClass | workers     | primary | failure         | failSecs | deploySecs | runSecs | nodeCleanSecs |
+      | ""         | "1-1"       | "1-1" | "0-0" | "powerstore" | "powerstore-nfs"   | "one-third" | "zero"  | "interfacedown" | 120      | 240        | 300     | 300           |
+      | ""         | "1-1"       | "1-1" | "0-0" | "powerstore" | "powerstore-iscsi"   | "one-third" | "zero"  | "interfacedown" | 120      | 240        | 300     | 300           |
+      #| ""         | "1-1"       | "1-1" | "0-0" | "powerstore" | "powerstore-nvmetcp"   | "one-third" | "zero"  | "interfacedown" | 120      | 240        | 300     | 300           |
```
[the remark this answers is not in the document]
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:612]: As we have to run e2e & longevity tests via csm-pipeline which doesn't support nvme-tcp, we have commented this for now. Manual testing with nvme-tcp has been done.

File: internal/monitor/features/integration.feature [hunk 20]
```diff
@@ -621,6 +747,33 @@ Feature: Integration Test
...
+      #Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 240      | 600        | 600     | 600          |
+      # Slightly more pods, increasing number of vols and devs
+      | ""         | "3-5"       | "1-1" | "0-0" | "powerstore" | "powerstore-nfs"   | "one-third" | "zero"  | "reboot" | 240      | 900        | 900     | 900           |
+      #Small number of pods, increasing number of vols and devs
+      | ""         | "1-2"       | "1-1" | "0-0" | "powerstore" | "powerstore-iscsi"   | "one-third" | "zero"  | "reboot" | 240      | 600        | 600     | 600          |
+      # Slightly more pods, increasing number of vols and devs
+      | ""         | "3-5"       | "1-1" | "0-0" | "powerstore" | "powerstore-iscsi"   | "one-third" | "zero"  | "reboot" | 240      | 900        | 900     | 900           |
+      #Small number of pods, increasing number of vols and devs
+      #| ""         | "1-2"       | "1-1" | "0-0" | "powerstore" | "powerstore-nvmetcp"   | "one-third" | "zero"  | "reboot" | 240      | 600        | 600     | 600          |
```
[the remark this answers is not in the document]
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:774]: As we have to run e2e & longevity tests via csm-pipeline which doesn't support nvme-tcp, we have commented this for now. Manual testing with nvme-tcp has been done.

File: internal/monitor/features/integration.feature (hunk 12 above)
Comment (suryagupta4) [reply internal/monitor/features/integration.feature:831]: As we have to run e2e & longevity tests via csm-pipeline which doesn't support nvme-tcp, we have commented this for now. Manual testing with nvme-tcp has been done.

File: test/sh/scaleup-powerstore.sh (hunk 16 above)
Comment (suryagupta4) [reply test/sh/scaleup-powerstore.sh]: updated

File: test/podmontest/insps.sh (hunk 15 above)
Comment (suryagupta4) [reply test/podmontest/insps.sh]: updated

File: test/podmontest/deploy/values-powerstore-nvme.yaml (hunk 14 above)
Comment (suryagupta4) [reply test/podmontest/deploy/values-powerstore-nvme.yaml:2]: We updated this with reference to Isilon changes.
https://github.com/dell/karavi-resiliency/blob/main/test/podmontest/deploy/values-isilon.yaml

File: test/podmontest/deploy/values-powerstore-nfs.yaml (hunk 13 above)
Comment (suryagupta4) [reply test/podmontest/deploy/values-powerstore-nfs.yaml:2]: We updated this with reference to Isilon changes.
https://github.com/dell/karavi-resiliency/blob/main/test/podmontest/deploy/values-isilon.yaml

**Review by shaynafinocchiaro — DISMISSED** [review]

**Review by gallacher — APPROVED** [review]

**Review by alikdell — DISMISSED** [review]

**Review by shaynafinocchiaro — APPROVED** [review]

**Review by alikdell — APPROVED** [review]

**Review by AkshaySainiDell — APPROVED** [review]

**Review by shaynafinocchiaro — APPROVED** [review]

**Pull request merged**